File: agentic-clj/src/agentic/trace.clj
(ns agentic.trace
  "Phase 7: observability — a minimal tracing SPI, one span per turn / timer fire / CEP match. The
   noop tracer costs nothing; the recording tracer captures completed spans for tests and local
   inspection; an OpenTelemetry exporter is an opt-in adapter (the heavy dependency stays out of the
   core). Mirror of jagentic-core's org.jagentic.core.trace.* — behaviour at parity, expressed
   idiomatically with protocol+deftype (matching agentic.timers).

   A Span carries key/value attributes and named events, then is ended. attr/event are chainable
   (they return the span). A Tracer starts named spans.")

;; ---- protocols ----

(defprotocol Span
  "A single trace span — attach attributes and named events, then end it."
  (span-attr [_ k v] "Attach a string attribute; returns the span (chainable).")
  (span-event [_ n] "Record a named event; returns the span (chainable).")
  (span-end [_] "Close the span. Returns nil."))

(defprotocol Tracer
  "Begin a span by name; attributes/events are added via the returned Span, closed with span-end."
  (start [_ name] "Begin a span; returns a Span."))

;; ---- noop: spans do nothing ----

(deftype NoopSpan []
  Span
  (span-attr [this _ _] this)
  (span-event [this _] this)
  (span-end [_] nil))

(deftype NoopTracer [span]
  Tracer
  (start [_ _name] span))

(defn noop-tracer
  "A tracer whose spans do nothing — the zero-cost default."
  []
  (->NoopTracer (->NoopSpan)))

;; ---- recording: capture completed spans in end() order ----

;; A span mutates its attrs/events while open (an atom-backed map + vector), then on end appends an
;; immutable {:name :attrs :events} snapshot to the tracer's recorded vector.

(deftype RecordingSpan [name !attrs !events !recorded]
  Span
  (span-attr [this k v]
    (swap! !attrs assoc k v)
    this)
  (span-event [this n]
    (swap! !events conj n)
    this)
  (span-end [_]
    (swap! !recorded conj {:name name :attrs @!attrs :events @!events})
    nil))

(deftype RecordingTracer [!recorded]
  Tracer
  (start [_ name]
    (->RecordingSpan name (atom {}) (atom []) !recorded)))

(defn recording-tracer
  "A tracer that records completed spans in end() order — for tests and local inspection."
  []
  (->RecordingTracer (atom [])))

(defn spans
  "The completed spans of a recording tracer, in end() order: a vector of {:name :attrs :events}."
  [^RecordingTracer tracer]
  @(.-!recorded tracer))

(defn names
  "Span names of a recording tracer, in end() order."
  [tracer]
  (mapv :name (spans tracer)))

File: ports/pyagentic/pyagentic/trace.py
"""Phase 7: observability — a minimal, idiomatic tracing SPI.

A span per turn, timer fire, and CEP match. The :data:`NOOP` default costs nothing;
:class:`RecordingTracer` captures completed spans for tests and local inspection; an
OpenTelemetry exporter is an opt-in adapter (the heavy dependency stays out of the
core). Engine adapters can bridge to their native tracing behind the same SPI.

The portable counterpart of the Java ``org.jagentic.core.trace`` SPI — same shapes,
same semantics: ``attr``/``event`` are chainable, ``end()`` closes the span, and a
RecordingTracer records spans in ``end()`` order.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from threading import Lock
from typing import Dict, List, Protocol


class Span(Protocol):
    """A single trace span — attach key/value attributes and named events, then
    :meth:`end`. ``attr`` and ``event`` return the span so they chain."""

    def attr(self, key: str, value: str) -> "Span":
        """Set an attribute; returns ``self`` for chaining."""
        ...

    def event(self, name: str) -> "Span":
        """Record a named event; returns ``self`` for chaining."""
        ...

    def end(self) -> None:
        """Close the span."""
        ...


class _NoopSpan:
    """A span that drops everything — the zero-cost default."""

    def attr(self, key: str, value: str) -> "Span":
        return self

    def event(self, name: str) -> "Span":
        return self

    def end(self) -> None:
        pass


class Tracer(Protocol):
    """Minimal tracing SPI: begin a span; attributes/events are added via the returned
    :class:`Span`, closed with ``end()``."""

    def start(self, name: str) -> Span:
        """Begin a span named ``name``."""
        ...


class NoopTracer:
    """A tracer whose spans are all no-ops — the default."""

    def start(self, name: str) -> Span:
        return _NOOP_SPAN


# Module-level singletons: the shared no-op span and tracer.
_NOOP_SPAN: Span = _NoopSpan()
NOOP: Tracer = NoopTracer()


@dataclass
class Recorded:
    """One completed span, as captured by :class:`RecordingTracer`."""

    name: str
    attrs: Dict[str, str] = field(default_factory=dict)
    events: List[str] = field(default_factory=list)


class _RecordingSpan:
    """A span that, on :meth:`end`, appends a :class:`Recorded` to its tracer."""

    def __init__(self, tracer: "RecordingTracer", name: str) -> None:
        self._tracer = tracer
        self._name = name
        self._attrs: Dict[str, str] = {}
        self._events: List[str] = []

    def attr(self, key: str, value: str) -> "Span":
        self._attrs[key] = value
        return self

    def event(self, name: str) -> "Span":
        self._events.append(name)
        return self

    def end(self) -> None:
        self._tracer._record(Recorded(self._name, self._attrs, self._events))


class RecordingTracer:
    """A tracer that records completed spans in ``end()`` order — for tests and local
    inspection. Thread-safe."""

    def __init__(self) -> None:
        self._spans: List[Recorded] = []
        self._guard = Lock()

    def start(self, name: str) -> Span:
        return _RecordingSpan(self, name)

    def _record(self, recorded: Recorded) -> None:
        with self._guard:
            self._spans.append(recorded)

    def spans(self) -> List[Recorded]:
        """Completed spans in ``end()`` order (a copy)."""
        with self._guard:
            return list(self._spans)

    def names(self) -> List[str]:
        """Span names in ``end()`` order."""
        return [r.name for r in self.spans()]
